### PR TITLE
Added support for custom domain names for Designer and Client interfaces

### PR DESCRIPTION
**Pull request by rajc-aws**
Updated QnABot to support use of custom domain names for Designer and Client user interfaces. 

Following files were added/updated:

Added: 
- docs/custom_domain_name_setup/README.md
- docs/custom_domain_name_setup/deploy_api_action.png
- docs/custom_domain_name_setup/stage_variables.png

Updated: 
- lambda/translate/index.js
- templates/master/routes/bot/get.resp.vm
- templates/master/routes/examples/info.vm
- templates/master/routes/examples/list.vm
- templates/master/routes/examples/photos.vm
- templates/master/routes/jobs/info.vm
- templates/master/routes/jobs/list-export.vm
- templates/master/routes/jobs/list-testall.vm
- templates/master/routes/jobs/list.vm
- templates/master/routes/root/info.vm
- website/js/lib/store/api/actions/index.js

Commit references:
- https://github.com/aws-solutions/aws-qnabot/commit/327d62185b3c041799fc493401dc12d47cb2a006
- https://github.com/aws-solutions/aws-qnabot/commit/c6bb19b34e327bf9acc55f14d8fc9f24be8a1307
- https://github.com/aws-solutions/aws-qnabot/commit/155de69381a82fcaadfcf91b8601b1e61fa32b8d
- https://github.com/aws-solutions/aws-qnabot/commit/14ad252a3fcefa5c65bc979ba429eb8088396733
- https://github.com/aws-solutions/aws-qnabot/commit/c253e801b8e4aed77c578124a664566441fa1a44

Submitting this pull request for review and next steps.

### CODE REVIEW

**Review by rstrahan — COMMENTED** [review]
Thanks Raj!  Can you add a section to discuss any implications to subsequent updates of QnABot after doing all this customization. Will the customizations be overwritten by updates? (I'm thinking 'yes' if the updates affect API gateway or Cognito resources, but probably not otherwise?).  Ideally QnAbot would be able to automate / preserve custom domain name configuration by accepting an (optional) custom domain name parameter in the CF template, and using it in the API Gateway / Cognito resources as needed to avoid breaking custom setups?
Other than upgrade concerns, as long as it is independently tested, i think it's great! I have not myself tested the steps, so let's make sure we have someone lined up to do the testing.
Thanks!

File: docs/custom_domain_name_setup/README.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+Create two mappings:
+
+##### Mapping 1: 
+- `API` = select the QnABot deployment you would like to use. The QnABot API will be named with the same name as the CloudFormation Stack name you used when you deployed the AWS QnABot solution
+- `Stage` = `prod`	- this is the default stage created for the QnABot deployment
+
+##### Mapping 2: 
+- `API` = select the QnABot deployment you would like to use. The QnABot API will be named with the same name as the CloudFormation Stack name you used when you deployed the AWS QnABot solution
+- `Stage` = `prod`	- this is the default stage created for the QnABot deployment
+- `Path` = `prod`	- for routing requests
```
Comment (tabdunabi) [inline docs/custom_domain_name_setup/README.md:30]: `Mapping 2` is exactly the same as `Mapping 1`, except the `Path` part?

File: templates/master/routes/examples/list.vm [hunk 2]
```diff
@@ -1,4 +1,4 @@
-#set ( $root="https://${!context.apiId}.execute-api.${AWS::Region}.amazonaws.com/${!context.stage}")
+#set ($root="https://${!context.domainName}/${!context.stage}")
```
Comment (tabdunabi) [inline templates/master/routes/examples/list.vm:1]: How this world work for the default API url (`$root="https://${!context.apiId}.execute-api.${AWS::Region}.amazonaws.com/${!context.stage}"`)?, where did you set `context.domainName` to have the correct value?

File: docs/custom_domain_name_setup/README.md (hunk 1 above)
Comment (rajc-aws) [reply docs/custom_domain_name_setup/README.md:30]: Thats correct. There are some API requests that require the path and some dont - being the case 2 mappings are used. 

API Gateway routes requests to the API mapping that has the longest matching path. API Gateway considers only the paths configured for API mappings, and not API routes, to select the API to invoke. If no path matches the request, API Gateway sends the request to the API that you've mapped to the empty path (none).

**Review by tabdunabi — APPROVED** [review]
LGTM.

**Pull request merged**